### PR TITLE
Jetpack Boost: Portable DataSync

### PR DESCRIPTION
**Pull request by pyronaur**
Refactoring DataSync for easier portability to other approaches.

## Proposed changes:

Bundle the core of data sync closer together - this is still separate enough for it to be used in various shapes, but a lot easier to boot up.

### Other information:

- [ ] Have you written new tests for your changes, if applicable?
- [ ] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [ ] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Testing instructions:
Test everything in the Jetpack Boost dashboard that makes a request

### CODE REVIEW

File: projects/js-packages/svelte-data-sync-client/src/initializeClient.ts [hunk 1]
```diff
@@ -89,25 +25,25 @@ function setupRestApi( namespace: string ) {
...
-		const syncedStore = new SyncedStore< z.infer< T > >( value );
-		const endpoint = new API_Endpoint< z.infer< T > >( api, valueName, schema );
+		const syncedStore = new SyncedStore< Value >( endpoint.getInitialValue() );
+
+		// The client doesn't need the whole store object.
+		// Only expose selected public methods:
+		const store = syncedStore.getPublicInterface();
 
 		/**
 		 * Wire up store to the endpoint.
```
Comment (dilirity) [inline projects/js-packages/svelte-data-sync-client/src/initializeClient.ts:49]: Comment needs an update - nonce is handled elsewhere now.

File: projects/js-packages/svelte-data-sync-client/src/DataSync.ts [hunk 2]
```diff
@@ -0,0 +1,286 @@
...
+		const result = await this.attemptRequest( url, args );
+
+		let data;
+		const text = await result.text();
+		try {
+			data = JSON.parse( text );
+		} catch ( e ) {
+			// eslint-disable-next-line no-console
+			console.error( 'Failed to parse the response\n', { url, text, result, error: e } );
+			throw new ApiError( url, 'json_parse_error', 'Failed to parse the response' );
```
Comment (dilirity) [inline projects/js-packages/svelte-data-sync-client/src/DataSync.ts:191]: While going up the chain where request is used, I noticed there's no try catch for errors like this one.

**Review by dilirity — CHANGES_REQUESTED** [review]
Validation is dead. Long live parsin :D

File: projects/js-packages/svelte-data-sync-client/src/initializeClient.ts (hunk 1 above)
Comment (pyronaur) [reply projects/js-packages/svelte-data-sync-client/src/initializeClient.ts:49]: Nice catch! 👍

File: projects/js-packages/svelte-data-sync-client/src/DataSync.ts (hunk 2 above)
Comment (pyronaur) [reply projects/js-packages/svelte-data-sync-client/src/DataSync.ts:191]: When a request fails - it should throw an error, whoever is triggering the method should catch it.

Most of the time it's going to be caught by SyncedStore.ts to retry the request.

**Comment by pyronaur** [comment]
> Validation is dead. Long live parsin :D

Exactly!
https://lexi-lambda.github.io/blog/2019/11/05/parse-don-t-validate/

File: projects/js-packages/svelte-data-sync-client/src/DataSync.ts (hunk 2 above)
Comment (dilirity) [reply projects/js-packages/svelte-data-sync-client/src/DataSync.ts:191]: Ah okay - thank you! Missed that 😅

**Review by dilirity — APPROVED** [review]
:shipit:

**Pull request merged**